### ISSUE TITLE
Hard links in shadow changes without modification on alter

### ISSUE THREAD

**Issue by mallniya** [ClickHouse/ClickHouse#24201]
**Describe the unexpected behaviour**
I have simple backup routine, proposed in [official documentation](https://clickhouse.tech/docs/en/operations/backup/): through `ALTER TABLE ... FREEZE PARTITION ...` and subsequent `tar`ing of `shadow` directory. Though it works at large, one implication raises in between `tar` and hardlinks in `shadow`: `tar` is sensible for on-the-fly file metadata changes, which are changed when altering tables, standing before these hardlinks.

**How to reproduce**
* Version 21.3.2.5
* `CREATE TABLE` statements for all tables involved:
```
CREATE TABLE default.iampartitioned
(
    `date` Date
)
ENGINE = MergeTree
PARTITION BY toYYYYMM(date)
ORDER BY date
SETTINGS index_granularity = 8192
```
with random dates: `insert into iampartitioned values ('2021-02-23');`
* Queries to run that lead to unexpected result:
Generate shadow with `ALTER TABLE iampartitioned FREEZE;` and check change time of created hardlinks:
```
# stat shadow/1/data/default/iampartitioned/*/*.bin | grep Change
Change: 2021-05-17 18:01:46.390883707 +0300
Change: 2021-05-17 18:01:46.391883667 +0300
Change: 2021-05-17 18:01:46.392883626 +0300
Change: 2021-05-17 18:01:46.394883545 +0300
```
At this point `tar` could be started to pack these hardlinks. if alter to table happens:
`alter table iampartitioned delete where date = '2021-02-23';`
then metadata changes (change time at least):
```
# stat shadow/1/data/default/iampartitioned/*/*.bin | grep Change
Change: 2021-05-17 18:03:09.025543573 +0300
Change: 2021-05-17 18:03:09.022543694 +0300
Change: 2021-05-17 18:03:09.029543411 +0300
Change: 2021-05-17 18:01:46.394883545 +0300
```
But I see no other differences (owner, mode, modification time) with `stat` before and after delete:
```
# stat shadow/1/data/default/iampartitioned/*/*.bin | grep Modify
Modify: 2021-02-20 18:01:10.105624719 +0300
Modify: 2021-02-21 05:52:19.206313638 +0300
Modify: 2021-05-17 18:00:08.322847289 +0300
Modify: 2021-05-17 18:00:15.545555385 +0300
```
So it seems, like some sort of internal clickhouse twitching causing hardlinks metadata to fluctuate. And if I'm unlucky enough (and with steady rate of deletions this is _always_) my `tar` fails with `tar: shadow/1/data/default/iampartitioned/202102_2_5_1_6/date.bin: file changed as we read it`.

**Expected behavior**
Frozen hardlinks to not change, as documentation states:
> The created copies of files are not handled by ClickHouse server

in other words: whatever happens within clickhouse (except, of course, `FREEZE` query), it doesn't affect shadow metadata.

**Comment by filimonov** [comment]
Everything is ok. It's linux changes the `Change` time once the original file get removed. 

```sh
## create some file 
$ cd /tmp
$ touch a

## and hardlink to it
$ ln a b 

## check stat of both
$ stat a
  File: a
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: fd01h/64769d	Inode: 46137357    Links: 2
Access: (0664/-rw-rw-r--)  Uid: ( 1000/mfilimonov)   Gid: ( 1000/mfilimonov)
Access: 2021-05-17 18:15:33.056115167 +0200
Modify: 2021-05-17 18:15:33.056115167 +0200
Change: 2021-05-17 18:15:43.356165730 +0200
 Birth: -

$ stat b
  File: b
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: fd01h/64769d	Inode: 46137357    Links: 2
Access: (0664/-rw-rw-r--)  Uid: ( 1000/mfilimonov)   Gid: ( 1000/mfilimonov)
Access: 2021-05-17 18:15:33.056115167 +0200
Modify: 2021-05-17 18:15:33.056115167 +0200
Change: 2021-05-17 18:15:43.356165730 +0200
 Birth: -

## till now everything is in sync.
## now let's unlink a

$ rm a

## and now you can see that Change time of the file (for b) was updated 

$ stat b
  File: b
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: fd01h/64769d	Inode: 46137357    Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1000/mfilimonov)   Gid: ( 1000/mfilimonov)
Access: 2021-05-17 18:15:33.056115167 +0200
Modify: 2021-05-17 18:15:33.056115167 +0200
Change: 2021-05-17 18:16:00.156248548 +0200
 Birth: -
```

**Comment by filimonov** [comment]
So just ignore that field. Everything is ok, nothing to worry about.

**Comment by mallniya** [comment]
I can ignore it, but `tar` can't (it could not show the warning, but it still would fail with rc == 1).
At least I know "why?" now, thanks.

**Comment by filimonov** [comment]
> but `tar` can't (it could not show the warning, but it still would fail with rc == 1).

Tar has plenty of options to fine-tune it. 
Please share the step you do and the errors you get.

**Comment by mallniya** [comment]
That's nice of you to help.
I have the following pipe: `set -o pipefail ; tar -c metadata/default/iampartitioned.sql metadata/default.sql shadow/*/data/default/iampartitioned | /usr/bin/zstd -T5 | ...`  and it fails with rc == 1 with hardlink changes. `--warning=no-file-changed` is no use here -- it only masks the `file changed as we read it`, not preventing tar of failing.

**Comment by filimonov** [comment]
> That's nice of you to help.
> I have the following pipe: `set -o pipefail ; tar -c metadata/default/iampartitioned.sql metadata/default.sql shadow/*/data/default/iampartitioned | /usr/bin/zstd -T5 | ...` and it fails with rc == 1 with hardlink changes. `--warning=no-file-changed` is no use here -- it only masks the `file changed as we read it`, not preventing tar of failing.

But do it really fail? Or it's only a matter or bad result code, while the archive is correct?

**Comment by filimonov** [comment]
Also you can try to use PIPESTATUS instead of pipefail
https://unix.stackexchange.com/a/14276

Or try to pass attributes the command line (not sure if it will work) or pass the files to the RAR through stdin - see some ideas in
https://unix.stackexchange.com/a/391502

**Comment by mallniya** [comment]
> But do it really fail? Or it's only a matter or bad result code, while the archive is correct?

Yes, you're right, I can verify it only now on live experiments (as oppose to cron logs) without pipe.
>Exit code 1 for create command literally means "some files was changed", and it's ok to ignore it in that case

Who would have thought (without any irony), what a coincidence that exactly this case is exit code 1. My tar 1.26 man doesn't contain this valuable info.
>See https://stackoverflow.com/a/24012292/1555175

There is case without pipe, which is required in my use case (I can't use intermediate local archive, it's too big), which brings us to the:
>Also you can try to use PIPESTATUS instead of pipefail

this should work, though complicating already tangled code.
>Or try to pass attributes the command line

I already did try `--mode`, no luck.
>pass the files to the tar through stdin

`find ... | tar --no-recursion -T - -c | ...` doesn't change a situation.
